### PR TITLE
[ValueTracking] Merge `cannotBeOrderedLessThanZeroImpl` into `computeKnownFPClass`

### PR DESCRIPTION
**Pull request by dtcxzyw**
This patch merges the logic of `cannotBeOrderedLessThanZeroImpl` into `computeKnownFPClass` to improve the signbit inference.

### CODE REVIEW

File: llvm/lib/Analysis/InstructionSimplify.cpp [hunk 1]
```diff
@@ -5740,9 +5740,9 @@ static Value *simplifyFMAFMul(Value *Op0, Value *Op1, FastMathFlags FMF,
       return ConstantFP::getZero(Op0->getType());
 
     // +normal number * (-)0.0 --> (-)0.0
-    if (isKnownNeverInfOrNaN(Op0, Q.DL, Q.TLI, 0, Q.AC, Q.CxtI, Q.DT) &&
-        // TODO: Check SignBit from computeKnownFPClass when it's more complete.
-        SignBitMustBeZero(Op0, Q.DL, Q.TLI))
+    KnownFPClass Known = computeKnownFPClass(
+        Op0, FMF, Q.DL, fcInf | fcNan, /*Depth*/ 0, Q.TLI, Q.AC, Q.CxtI, Q.DT);
```
Comment (arsenm) [inline llvm/lib/Analysis/InstructionSimplify.cpp:5744]: /*Depth=/* so clang-format recognizes it

File: llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp [hunk 2]
```diff
@@ -2408,19 +2408,19 @@ Instruction *InstCombinerImpl::visitCallInst(CallInst &CI) {
...
-      Value *Fabs = Builder.CreateUnaryIntrinsic(Intrinsic::fabs, Mag, II);
-      return replaceInstUsesWith(*II, Builder.CreateFNegFMF(Fabs, II));
+    if (std::optional<bool> KnownSignBit = computeKnownFPSignBit(
+            Sign, getDataLayout(), &TLI, /*Depth*/ 0, &AC, II, &DT)) {
+      if (*KnownSignBit) {
+        // If we know that the sign argument is negative, reduce to FNABS:
+        // copysign Mag, -Sign --> fneg (fabs Mag)
+        Value *Fabs = Builder.CreateUnaryIntrinsic(Intrinsic::fabs, Mag, II);
+        return replaceInstUsesWith(*II, Builder.CreateFNegFMF(Fabs, II));
+      } else {
```
Comment (arsenm) [inline llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp:2418]: No else after return

File: llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp [hunk 3]
```diff
@@ -2408,19 +2408,19 @@ Instruction *InstCombinerImpl::visitCallInst(CallInst &CI) {
...
-    }
-    // TODO: There should be a ValueTracking sibling like SignBitMustBeOne.
-    const APFloat *C;
-    if (match(Sign, m_APFloat(C)) && C->isNegative()) {
-      // If we know that the sign argument is negative, reduce to FNABS:
-      // copysign Mag, -Sign --> fneg (fabs Mag)
-      Value *Fabs = Builder.CreateUnaryIntrinsic(Intrinsic::fabs, Mag, II);
-      return replaceInstUsesWith(*II, Builder.CreateFNegFMF(Fabs, II));
+    if (std::optional<bool> KnownSignBit = computeKnownFPSignBit(
+            Sign, getDataLayout(), &TLI, /*Depth*/ 0, &AC, II, &DT)) {
```
Comment (arsenm) [inline llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp:2412]: /*Depth=*/

File: llvm/lib/Analysis/ValueTracking.cpp [hunk 4]
```diff
@@ -3737,205 +3737,6 @@ Intrinsic::ID llvm::getIntrinsicForCallSite(const CallBase &CB,
...
-      //   pow(-0, exp)   --> -0 if exp is positive odd.
-      //   pow(-inf, exp) --> -0 if exp is negative odd.
-      //   pow(-inf, exp) --> -inf if exp is positive odd.
-      //
-      // Therefore, if !SignBitOnly, we can return true if x >= +0 or x is NaN,
-      // but we must return false if x == -0.  Unfortunately we do not currently
-      // have a way of expressing this constraint.  See details in
-      // https://llvm.org/bugs/show_bug.cgi?id=31702.
-      return cannotBeOrderedLessThanZeroImpl(I->getOperand(0), DL, TLI,
-                                             SignBitOnly, Depth + 1);
```
Comment (arsenm) [inline llvm/lib/Analysis/ValueTracking.cpp:3918]: I think a few too many things are happening at once in this patch. Most significantly, this TODO about being incorrect seems inadequately tested. I would hope to see some powi test changes showing this bug has been fixed

Comment (dtcxzyw) [reply llvm/lib/Analysis/ValueTracking.cpp:3885]: Done.

**Comment by dtcxzyw** [comment]
Ping.

**Comment by dtcxzyw** [comment]
Ping.

**Review by arsenm — APPROVED** [review]

**Pull request merged**